backend/rates/rates.go

```diff
@@ -20,60 +20,124 @@ import (
 
-// RateUpdater implements coin.RateUpdater.
+type exchangeRate struct {
+	timestamp time.Time
+	value     float64
+}
+
+// RateUpdater provides cryptocurrency-to-fiat conversion rates.
 type RateUpdater struct {
 	observable.Implementation
-	last       map[string]map[string]float64
+
+	httpClient *http.Client
 	log        *logrus.Entry
-	socksProxy socksproxy.SocksProxy
+
+	// last contains most recent conversion to fiat, keyed by a coin.
+	last map[string]map[string]float64
+
+	historyMu sync.RWMutex
+	// history contains historical conversion rates in asc order, keyed by a coin.
+	// TODO: store in bolt DB
+	// TODO: support different fiat currencies? it's in USD at the moment
+	history map[string][]exchangeRate
+
+	// CoinGecko is where updater gets the historical conversion rates.
+	// See https://www.coingecko.com/en/api for details.
+	coingeckoURL string
 }
 
 // NewRateUpdater returns a new rates updater.
-func NewRateUpdater(socksProxy socksproxy.SocksProxy) *RateUpdater {
-	ratesUpdater := &RateUpdater{
-		last:       map[string]map[string]float64{},
-		log:        logging.Get().WithGroup("rates"),
-		socksProxy: socksProxy,
+func NewRateUpdater(client *http.Client) *RateUpdater {
+	return &RateUpdater{
+		last:         make(map[string]map[string]float64),
+		history:      make(map[string][]exchangeRate),
+		log:          logging.Get().WithGroup("rates"),
+		httpClient:   client,
+		coingeckoURL: coingeckoAPIV3,
 	}
-	go ratesUpdater.start()
-	return ratesUpdater
 }
 
-// Last returns the last rates for a given coin and fiat or nil if not available.
+// Last returns the most recent conversion rates.
+// The returned map is keyed by a crypto coin with values mapped by fiat rates.
+// RateUpdater assumes the returned value is never modified by the callers.
 func (updater *RateUpdater) Last() map[string]map[string]float64 {
 	return updater.last
 }
 
-func (updater *RateUpdater) update() {
-	client, err := updater.socksProxy.GetHTTPClient()
-	if err != nil {
-		updater.log.Printf("Error getting http client %v\n", err)
-		updater.last = nil
-		return
+// PriceUSD returns a historical exchange rate for the given coin.
+// The returned value may be imprecise if at arg matches no timestamp exactly.
+// In this case, linear interpolation is used as an approximation.
+// If no data is available with the given args, PriceUSD returns 0.
+//nolint:staticcheck
+func (updater *RateUpdater) PriceUSD(coin string, at time.Time) float64 {
+	coin = "BTC" // TODO: support all coins; remove nolint:staticcheck
+	updater.historyMu.RLock()
+	defer updater.historyMu.RUnlock()
+	data := updater.history[coin]
+	// Find an index of the first entry older or equal the at timestamp.
+	idx := sort.Search(len(data), func(i int) bool {
+		return !data[i].timestamp.Before(at)
+	})
+	if idx == len(data) {
+		return 0 // no data
+	}
+	if idx == 0 || data[idx].timestamp.Equal(at) {
+		return data[idx].value // can't or don't need to interpolate
```

Comment:
with `idx==0`, the value in the past for which we do not have rates is very likely totally wrong. better return `0` in this case (no data).

E.g. the price in the year 2010 should not be USD 10'000 just because we only have the last three months available ;)

Reply:
For the record, this is the only "semantic" issue in the PR, the rest are nits or things than can be improved later.

Reply:
true. better return 0. however, `idx == 0` **and** equal timestamps is still a good case to return the non-zero value.